install/scripts/docker_entrypoint.sh

```diff
@@ -16,97 +16,168 @@
-# TODO switch debugging on/off via env var
-gsutil cp ${BUCKET}/configs/forseti_conf_server.yaml /forseti-security/configs/forseti_conf_server.yaml
-gsutil cp -r ${BUCKET}/rules /forseti-security/
-
-
-# TODO Error handling for gsutil cp
-
-# Start Forseti server
-# This requires the cloud sql proxy (side car) container is running on 127.0.0.1:3306
-
-# TODO switch debugging on/off via env var
-forseti_server \
---endpoint "localhost:50051" \
---forseti_db "mysql://root@${CLOUDSQLPROXY_SERVICE_HOST}:${CLOUDSQLPROXY_SERVICE_PORT}/forseti_security" \
---services scanner model inventory explain notifier \
---config_file_path "/forseti-security/configs/forseti_conf_server.yaml" \
---log_level=debug &
-#--enable_console_log
-
-# Below cut and paste from run_forseti.sh ######################################
-# Ideally just call run_forseti.sh directly but for now its not quite right for us in GKE
-
-# Wait until the service is started
-sleep 10s
-
-# Set the output format to json
-forseti config format json
-
-# Purge inventory.
-# Use retention_days from configuration yaml file.
-forseti inventory purge
-
-# Run inventory command
-MODEL_NAME=$(/bin/date -u +%Y%m%dT%H%M%S)
-echo "Running Forseti inventory."
-forseti inventory create --import_as ${MODEL_NAME}
-echo "Finished running Forseti inventory."
-sleep 5s
-
-GET_MODEL_STATUS="forseti model get ${MODEL_NAME} | python -c \"import sys, json; print json.load(sys.stdin)['status']\""
-MODEL_STATUS=`eval $GET_MODEL_STATUS`
-
-if [ "$MODEL_STATUS" == "BROKEN" ]
-    then
-        echo "Model is broken, please contact discuss@forsetisecurity.org for support."
-        exit
-fi
-
-# Run model command
-echo "Using model ${MODEL_NAME} to run scanner"
-forseti model use ${MODEL_NAME}
-# Sometimes there's a lag between when the model
-# successfully saves to the database.
-sleep 10s
-echo "Forseti config: $(forseti config show)"
-
-# Run scanner command
-echo "Running Forseti scanner."
-scanner_command=`forseti scanner run`
-scanner_index_id=`echo ${scanner_command} | grep -o -P '(?<=(ID: )).*(?=is created)'`
-echo "Finished running Forseti scanner."
-sleep 10s
-
-# Run notifier command
-echo "Running Forseti notifier."
-# forseti notifier run --scanner_index_id ${scanner_index_id}
-forseti notifier run
-echo "Finished running Forseti notifier."
-sleep 10s
-
-# Clean up the model tables
-echo "Cleaning up model tables"
-forseti model delete ${MODEL_NAME}
-
-# End cut and paste from run_forseti.sh ######################################
+
+# TODO Error handling in all functions
+# For now, just stop the script if an error occurs
+set -e
+
+# Declare variables and set default values
+BUCKET=
+LOG_LEVEL=info
+SERVICES="scanner model inventory explain notifier"
+RUN_SERVER=true
+RUN_CLIENT=false
+
+# Note
+# CLOUDSQLPROXY_SERVICE_HOST
+# CLOUDSQLPROXY_SERVICE_PORT
+# are environment variables set by k8s
+# TODO process these as command line args if not running on k8s
+
+download_configuration_files(){
```

Comment:
what will be the service account here to access the configuration files? the compute engine default service account, or the service account specified by the `GOOGLE_APPLICATION_CREDENTIALS` environment variable?

Reply:
If its the default service account, then it might not have the authorization to access the bucket unless the bucket is in the same project. would it be good to do something like [this](https://cloud.google.com/sdk/gcloud/reference/auth/activate-service-account)?

Reply:
Forseti service account credentials json file added as k8s secret and mounted to the images.
Cloud SQL proxy refs the file directly; For Forseti we set the env variable GOOGLE_APPLICATION_CREDENTIALS to point to the mounted secret.

Reply:
We should not be using exported service account keys in general. Getting hold of the key file is equivalent to being able to anonymously assume the identity of that service account from any system. There is no current support binding between the service account and the k8s cluster.

Rather the requirement at this time is to run a node pool for forseti that has the node pool default service account set to the forseti service account, and use the existing default service account code to get the credential from metadata.

When per pod identity is supported in gke, then we should evaluate switching to that.

Reply:
> Forseti service account credentials json file added as k8s secret and mounted to the images.
> Cloud SQL proxy refs the file directly; For Forseti we set the env variable GOOGLE_APPLICATION_CREDENTIALS to point to the mounted secret.

yes, I'm asking because I just deployed a cluster using the GCP resources(bucket and credentials etc) from another project (didn't want to create a new set of duplicate resources). The gsutil command gave an error in the pod log saying the compute engine default service account cannot download the bucket config file. it seems that the gsutil command is called with the default service account, instead of the service account in the credentials json file

I then tried to create the bucket on this project by installing forseti, that fixed the issue.

Reply:
> We should not be using exported service account keys in general. Getting hold of the key file is equivalent to being able to anonymously assume the identity of that service account from any system. There is no current support binding between the service account and the k8s cluster.
> 
> Rather the requirement at this time is to run a node pool for forseti that has the node pool default service account set to the forseti service account, and use the existing default service account code to get the credential from metadata.
> 
> When per pod identity is supported in gke, then we should evaluate switching to that.

so if we are using the default service account instead, that means we need to run the iam assignment process on that service account before starting sql proxy and the forseti server, correct?

Reply:
Right.

1) Create service account
2) Configure IAM bindings for resources the service account needs access to
3) Create a new node pool in the cluster, associate it with service account from 1
4) Ensure forseti pods target the correct node pool, https://cloud.google.com/kubernetes-engine/docs/concepts/node-pools#deploying_services_to_specific_node_pools

Reply:
fyi my clients intention is to potentially run Forseti server on docker/k8s outside of GCP (on another cloud provider that shall not be named :open_mouth: ) They are aware of the need for secure practices around key management.

Reply:
yes we are aware of that. we have a few questions regarding how your organization wanna proceed with Forseti, messaged you via slack ;)